### PR TITLE
Improved OIDC compliance

### PR DESCRIPTION
**Pull request by lbalmaceda**
### Changes

This update improves the SDK support for OpenID Connect. In particular, it modifies the sign in verification phase by substituting backchannel based checks with id_token validation where possible.

### Testing

Please describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. If this library has unit and/or integration testing, tests should be added for new functionality and existing tests should complete without errors. 

- [x] This change adds test coverage
- [x] This change has been tested on the latest version of Java or why not

### Checklist

- [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [x] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
- [x] All existing and new tests complete without errors

### CODE REVIEW

File: src/main/java/com/auth0/SignatureVerifier.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            return JWT.decode(token);
+        } catch (JWTDecodeException e) {
+            throw new TokenValidationException("ID token could not be decoded", e);
+        }
+    }
+
+    DecodedJWT verifySignature(String token) throws TokenValidationException {
+        DecodedJWT decoded = decodeToken(token);
+        if (!this.acceptedAlgorithms.contains(decoded.getAlgorithm())) {
+            throw new TokenValidationException(String.format("Signature algorithm of \"%s\" is not supported. Expected the ID token to be signed with \"%s\".", decoded.getAlgorithm(), this.acceptedAlgorithms));
```
Comment (joshcanhelp) [inline src/main/java/com/auth0/SignatureVerifier.java]: Error message is taking `this.acceptedAlgorithms` but there should only be a single alg accepted. Misnamed var?

File: src/main/java/com/auth0/RequestProcessor.java [hunk 2]
```diff
@@ -83,73 +84,82 @@ AuthorizeUrl buildAuthorizeUrl(HttpServletRequest request, String redirectUri, S
...
+                if (!responseTypeList.contains(KEY_ID_TOKEN)) {
+                    //If already verified the front-channel token, don't verify it again.
+                    idToken = codeExchangeTokens.getIdToken();
+                    if (idToken != null) {
+                        tokenVerifier.verify(idToken, verifyOptions);
+                    }
+                }
             }
+        } catch (TokenValidationException e) {
+            throw new IdentityVerificationException(JWT_VERIFICATION_ERROR, "An error occurred while trying to verify the Id Token.", e);
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/RequestProcessor.java]: 👍 change is good, but will make a separate commit that includes updates to the expected message in the test.

File: src/main/java/com/auth0/RequestProcessor.java [hunk 3]
```diff
@@ -83,73 +84,82 @@ AuthorizeUrl buildAuthorizeUrl(HttpServletRequest request, String redirectUri, S
...
+        } catch (Auth0Exception e) {
+            throw new IdentityVerificationException(API_ERROR, "An error occurred while exchanging the Authorization Code for Auth0 Tokens.", e);
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/RequestProcessor.java]: 👍 change is good, but will make separate commit with this change that includes updates to the test.

File: src/main/java/com/auth0/RequestProcessor.java [hunk 4]
```diff
@@ -199,34 +209,21 @@ private Tokens exchangeCodeForTokens(String authorizationCode, String redirectUr
...
-        String idToken = latestTokens.getIdToken() != null ? latestTokens.getIdToken() : tokens.getIdToken();
-        String refreshToken = latestTokens.getRefreshToken() != null ? latestTokens.getRefreshToken() : tokens.getRefreshToken();
-        String type = latestTokens.getType() != null ? latestTokens.getType() : tokens.getType();
-        Long expiresIn = latestTokens.getExpiresIn() != null ? latestTokens.getExpiresIn() : tokens.getExpiresIn();
+    private Tokens mergeTokens(Tokens frontChannelTokens, Tokens codeExchangeTokens) {
+        //Prefer access token from the code exchange
+        String accessToken = codeExchangeTokens.getAccessToken() != null ? codeExchangeTokens.getAccessToken() : frontChannelTokens.getAccessToken();
+        //Prefer id token from the front-channel
+        String idToken = frontChannelTokens.getIdToken() != null ? frontChannelTokens.getIdToken() : codeExchangeTokens.getIdToken();
+        String refreshToken = frontChannelTokens.getRefreshToken() != null ? frontChannelTokens.getRefreshToken() : codeExchangeTokens.getRefreshToken();
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/RequestProcessor.java]: Updated to not consider front channel refresh tokens at all.

File: src/main/java/com/auth0/SignatureVerifier.java (hunk 1 above)
Comment (jimmyjames) [reply src/main/java/com/auth0/SignatureVerifier.java]: The concrete implementation of `SignatureVerifier` will set the `acceptedAlgorithms`. So, the message will be specific for the specific signature verification method being used - e.g., when using HS256, if the token is actually signed with RS256, the exception message will be `Signature algorithm of "HS256" is not supported. Expected the ID token to be signed with "RS256".` Conversely, if configured for RS256 but the token is actually signed with HS256, the message would be the opposite.

File: src/main/java/com/auth0/RequestProcessor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/RequestProcessor.java]: Updated in a new commit

File: src/main/java/com/auth0/RequestProcessor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/RequestProcessor.java]: Updated in a new commit

File: README.md [hunk 5]
```diff
@@ -89,41 +90,67 @@ That's it! You have authenticated the user using Auth0.
...
-### Builder Options
-By default, the **Code Grant** flow will be preferred over other flows. This is the most secure and recommended way, read more about it [here](https://auth0.com/docs/api-auth/grant/authorization-code). This means that if the response type contains `code` along with other types, Code Grant will still be preferred.
-
-You can change the authentication behavior to use **Implicit Grant** instead. To do this you'll need to check in your Applications's Settings on the [Dashboard](https://manage.auth0.com/#/applications) which Algorithm is used by the Server to sign the tokens. The default algorithm is `HS256`, but it can be changed to `RS256` in the "Advanced Settings" section on the "OAuth" tab. Below you'll find some configuration examples:
+### Builder options
+By default, this library will execute the [Open ID Connect](https://openid.net/specs/openid-connect-core-1_0-final.html) **Code Grant** flow and verify the ID token if received. The library expects to interact with an Application that is signing the tokens using the **HS256 symmetric algorithm** with the client secret being the shared secret. You can customize this behavior with the builder methods below.
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply README.md]: Updated to improve clarity.

File: README.md [hunk 6]
```diff
@@ -89,41 +90,67 @@ That's it! You have authenticated the user using Auth0.
...
-The `JwkProvider` returned from the `JwkProviderBuilder` it's cached and rate limited, check it's [repository](https://github.com/auth0/jwks-rsa-java) to learn how to customize it.
+or **Hybrid Grant** flow.
+
+```java
+AuthenticationController authController = AuthenticationController.newBuilder("domain", "clientId", "clientSecret")
+    .withResponseType("id_token code")
+    .build();
+```
+
+#### Limiting the time since last authentication
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply README.md]: Removed.

File: README.md [hunk 7]
```diff
@@ -89,41 +90,67 @@ That's it! You have authenticated the user using Auth0.
...
+    .withAuthenticationMaxAge(60 * 1)   //1 minute
+    .build();
+```
 
 
 ### Troubleshooting
 
-Once you have created the instance of the `AuthenticationController` you can enable HTTP logging for all Requests and Responses if you need to debug a specific endpoint. Keep in mind that this will log everything including sensitive information. Don't use it in production environment.
+#### Allowing a clock skew
+Depending on where your application is being executed on, there might be slight differences between the computer's clock and the Auth0 servers' clock. This difference could lead to reject a valid ID token and thus, unauthorized access. To prevent this type of errors the library allows a default of 60 seconds of clock skewing or leeway. To specify a value bigger than that:
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply README.md]: Updated.

File: README.md [hunk 8]
```diff
@@ -89,41 +90,67 @@ That's it! You have authenticated the user using Auth0.
...
     .build();
 ```
 
-#### Using Implicit Grant with RS256 algorithm.
+The `JwkProvider` returned from the `JwkProviderBuilder` it's cached and rate limited by default. Check it's [repository](https://github.com/auth0/jwks-rsa-java) to learn how to customize these options.
+
+#### Choosing a different Response Type
 
-The tokens are signed using the Private Key. To verify them, the **Public Key** certificate must be obtained from a trusted source like the `well-known.json` file, which can be located locally or hosted by a server. For this example, we will use the one Auth0 hosts for your application. We can obtain it using the application's domain:
+**Code Grant** flow will be preferred over other flows. This is the most secure and recommended way, read more about it [here](https://auth0.com/docs/flows/concepts/auth-code).
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply README.md]: Updated.

File: src/main/java/com/auth0/RequestProcessor.java [hunk 9]
```diff
@@ -83,73 +84,82 @@ AuthorizeUrl buildAuthorizeUrl(HttpServletRequest request, String redirectUri, S
...
 
-        Tokens tokens = tokensFromRequest(req);
+        Tokens frontChannelTokens = getFrontChannelTokens(req);
+        Tokens codeExchangeTokens = frontChannelTokens;
+        List<String> responseTypeList = getResponseType();
+
+        if (responseTypeList.contains(KEY_ID_TOKEN) && frontChannelTokens.getIdToken() == null) {
+            throw new InvalidRequestException(MISSING_ID_TOKEN, "Id Token is missing from the response.");
+        }
+        if (responseTypeList.contains(KEY_TOKEN) && frontChannelTokens.getAccessToken() == null) {
```
[the remark this answers is not in the document]
Comment (lbalmaceda) [reply src/main/java/com/auth0/RequestProcessor.java]: I assume you say that because using implicit grant flows with form-post is as safe as using code grant flows when we want to save that extra code-exchange network request. This has been like this for some time now. If we are going to do such change or deprecation I suggest we discuss this first and start a project to update the rest of the SDKs as well.

File: README.md [hunk 10]
```diff
@@ -89,41 +90,64 @@ That's it! You have authenticated the user using Auth0.
...
-By default, the **Code Grant** flow will be preferred over other flows. This is the most secure and recommended way, read more about it [here](https://auth0.com/docs/api-auth/grant/authorization-code). This means that if the response type contains `code` along with other types, Code Grant will still be preferred.
+### Builder options
+
+By default, this library will execute the [Open ID Connect](https://openid.net/specs/openid-connect-core-1_0-final.html) **Authorization Code Flow** and verify the ID token (if received) using the **HS256 symmetric algorithm**.
 
-You can change the authentication behavior to use **Implicit Grant** instead. To do this you'll need to check in your Applications's Settings on the [Dashboard](https://manage.auth0.com/#/applications) which Algorithm is used by the Server to sign the tokens. The default algorithm is `HS256`, but it can be changed to `RS256` in the "Advanced Settings" section on the "OAuth" tab. Below you'll find some configuration examples:
+#### Signing Algorithms
 
+The **HS256 symmetric algorithm** is the default expected signing algorithm. Tokens are verified using the client secret found in your Auth0 Application's settings.
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply README.md]: Yeah, that's probably a good addition.

File: README.md [hunk 11]
```diff
@@ -89,41 +90,64 @@ That's it! You have authenticated the user using Auth0.
...
 
-#### Using Implicit Grant with HS256 algorithm
+If your application is using the **RS256 asymmetric algorithm**, tokens are signed using a private key and verified using the public key associated with your Auth0 domain.
+For RS256, configure a `JwkProvider` for your Auth0 domain to enable retrieving the public key needed during the verification phase:
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply README.md]: "For using" reads a little odd, but it could be "If using"

File: src/main/java/com/auth0/IdTokenVerifier.java [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+        if (decoded.getExpiresAt() == null) {
+            throw new TokenValidationException("Expiration Time (exp) claim must be a number present in the ID token");
+        }
+
+        cal.setTime(decoded.getExpiresAt());
+        cal.add(Calendar.SECOND, leeway);
+        Date expDate = cal.getTime();
+
+        if (now.after(expDate)) {
+            throw new TokenValidationException(String.format("Expiration Time (exp) claim error in the ID token; current time %d is after expiration time %d", now.getTime() / 1000, expDate.getTime() / 1000));
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/IdTokenVerifier.java]: Updated in a separate commit to also update tests.

File: src/main/java/com/auth0/IdTokenVerifier.java [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+        if (decoded.getIssuedAt() == null) {
+            throw new TokenValidationException("Issued At (iat) claim must be a number present in the ID token");
+        }
+
+        cal.setTime(decoded.getIssuedAt());
+        cal.add(Calendar.SECOND, -1 * leeway);
+        Date iatDate = cal.getTime();
+
+        if (now.before(iatDate)) {
+            throw new TokenValidationException(String.format("Issued At (iat) claim error in the ID token; current time %d is before issued at time %d", now.getTime() / 1000, iatDate.getTime() / 1000));
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/IdTokenVerifier.java]: Updated in a separate commit to also update tests.

File: README.md [hunk 14]
```diff
@@ -89,41 +90,64 @@ That's it! You have authenticated the user using Auth0.
...
 ```java
-JwkProvider jwkProvider = new JwkProviderBuilder("domain").build();
 AuthenticationController authController = AuthenticationController.newBuilder("domain", "clientId", "clientSecret")
     .withResponseType("id_token")
-    .withJwkProvider(jwkProvider)
     .build();
 ```
 
-The `JwkProvider` returned from the `JwkProviderBuilder` it's cached and rate limited, check it's [repository](https://github.com/auth0/jwks-rsa-java) to learn how to customize it.
+To use the **[Hybrid Flow](https://auth0.com/docs/api-auth/grant/hybrid)**, specify `id_token code` as the response type:
```
Comment (joshcanhelp) [inline README.md]:
```suggestion
To use the **[Hybrid Flow](https://auth0.com/docs/api-auth/grant/hybrid)**, specify `code id_token` as the response type:
```

Teeny tiny but if we ever do discovery, this would need to match [`response_types_supported `](https://auth0.auth0.com/.well-known/openid-configuration).

File: src/main/java/com/auth0/AuthenticationController.java [hunk 15]
```diff
@@ -82,37 +92,83 @@ public Builder withJwkProvider(JwkProvider jwkProvider) {
...
+            } else if (responseType.contains(RESPONSE_TYPE_CODE)) {
+                // Old behavior: To maintain 100% retro-compatibility when
+                // no explicit algorithm is set by the user, we
+                // must skip ID Token signature check!
+                signatureVerifier = new AlgorithmNameVerifier();
+            } else {
+                signatureVerifier = new SymmetricSignatureVerifier(clientSecret);
+            }
+
+            IdTokenVerifier.Options verifyOptions = createIdTokenVerificationOptions(domain, clientId, signatureVerifier);
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply src/main/java/com/auth0/AuthenticationController.java]: Still feels wrong to me ...

File: src/main/java/com/auth0/IdTokenVerifier.java [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+        }
+
+
+        if (verifyOptions.nonce != null) {
+            String nonceClaim = decoded.getClaim(NONCE_CLAIM).asString();
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/IdTokenVerifier.java]: We're getting type safety with the `asString`, as it will return `null` if the backing JSON value isn't a string.

File: src/main/java/com/auth0/IdTokenVerifier.java [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+            if (isEmpty(nonceClaim)) {
+                throw new TokenValidationException("Nonce (nonce) claim must be a string present in the ID token");
+            }
+            if (!verifyOptions.nonce.equals(nonceClaim)) {
+                throw new TokenValidationException(String.format("Nonce (nonce) claim mismatch in the ID token; expected \"%s\", found \"%s\"", verifyOptions.nonce, nonceClaim));
+            }
+        }
+
+        if (audience.size() > 1) {
+            String azpClaim = decoded.getClaim(AZP_CLAIM).asString();
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/IdTokenVerifier.java]: We're getting type safety with the `asString`, as it will return `null` if the backing JSON value isn't a string.

File: src/main/java/com/auth0/IdTokenVerifier.java [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+            if (isEmpty(azpClaim)) {
+                throw new TokenValidationException("Authorized Party (azp) claim must be a string present in the ID token when Audience (aud) claim has multiple values");
+            }
+            if (!verifyOptions.audience.equals(azpClaim)) {
+                throw new TokenValidationException(String.format("Authorized Party (azp) claim mismatch in the ID token; expected \"%s\", found \"%s\"", verifyOptions.audience, azpClaim));
+            }
+        }
+
+        if (verifyOptions.maxAge != null) {
+            Date authTime = decoded.getClaim(AUTH_TIME_CLAIM).asDate();
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/IdTokenVerifier.java]: We don't need to type check since we're using the strongly typed `DecodedJwt`.

File: src/main/java/com/auth0/IdTokenVerifier.java [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+                throw new TokenValidationException("Authentication Time (auth_time) claim must be a number present in the ID token when Max Age (max_age) is specified");
+            }
+
+            cal.setTime(authTime);
+            cal.add(Calendar.SECOND, verifyOptions.maxAge);
+            cal.add(Calendar.SECOND, leeway);
+            Date authTimeDate = cal.getTime();
+
+            if (now.after(authTimeDate)) {
+                throw new TokenValidationException(String.format("Authentication Time (auth_time) claim in the ID token indicates that too much time has passed since the last end-user authentication. Current time %d is after last auth at %d", now.getTime() / 1000, authTimeDate.getTime() / 1000));
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/IdTokenVerifier.java]: Updated per suggestion in a separate commit.

File: src/main/java/com/auth0/RequestProcessor.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply src/main/java/com/auth0/RequestProcessor.java]: I understand if this is backwards-compat but, regardless of the response mode, the response type should not include `token`

File: src/main/java/com/auth0/AsymmetricSignatureVerifier.java [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.JWTVerifier;
+import com.auth0.jwt.algorithms.Algorithm;
+import com.auth0.jwt.interfaces.RSAKeyProvider;
+
+import java.security.interfaces.RSAPrivateKey;
+import java.security.interfaces.RSAPublicKey;
+
+@SuppressWarnings("unused")
+class AsymmetricSignatureVerifier extends SignatureVerifier {
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/AsymmetricSignatureVerifier.java]: I think that's probably jst java and the use of inner classes.

File: src/main/java/com/auth0/IdTokenVerifier.java [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+        if (isEmpty(decoded.getSubject())) {
+            throw new TokenValidationException("Subject (sub) claim must be a string present in the ID token");
+        }
+
+        final List<String> audience = decoded.getAudience();
+        if (audience == null) {
+            throw new TokenValidationException("Audience (aud) claim must be a string or array of strings present in the ID token");
+        }
+        if (!audience.contains(verifyOptions.audience)) {
+            throw new TokenValidationException(String.format("Audience (aud) claim mismatch in the ID token; expected \"%s\" but found \"%s\"", verifyOptions.audience, decoded.getAudience()));
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/IdTokenVerifier.java]: Yes, but in the case of Java with java-jwt, the Audience claims are always a list. Even if we tried to assume the underlying type by inspecting the size of the list, we really can't be sure it wasn't just a list with one element in the actual claims. In this case we're going to have to diverge a bit from our reference example.

File: src/main/java/com/auth0/IdTokenVerifier.java [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+        //1 presence
+        if (isEmpty(token)) {
+            throw new TokenValidationException("ID token is required but missing");
+        }
+
+        //2 and 3
+        DecodedJWT decoded = verifyOptions.verifier.verifySignature(token);
+
+        //4. Claims
+        if (isEmpty(decoded.getIssuer())) {
```
[the remark this answers is not in the document]
Comment (lbalmaceda) [reply src/main/java/com/auth0/IdTokenVerifier.java]: the java-jwt dependency takes care of checking types for us. Will return null if the value is not a string; which is checked alongside with empty strings in the `isEmpty` function.

File: src/main/java/com/auth0/IdTokenVerifier.java [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+
+        if (now.before(iatDate)) {
+            throw new TokenValidationException(String.format("Issued At (iat) claim error in the ID token; current time %d is before issued at time %d", now.getTime() / 1000, iatDate.getTime() / 1000));
+        }
+
+
+        if (verifyOptions.nonce != null) {
+            String nonceClaim = decoded.getClaim(NONCE_CLAIM).asString();
+            if (isEmpty(nonceClaim)) {
+                throw new TokenValidationException("Nonce (nonce) claim must be a string present in the ID token");
```
[the remark this answers is not in the document]
Comment (jimmyjames) [reply src/main/java/com/auth0/IdTokenVerifier.java]: I think that might make the code a bit too noisy... if Java had better String interpolation it wouldn't be so bad, but I think it's ok as-is.

File: README.md (hunk 14 above)
Comment (lbalmaceda) [reply README.md]: https://openid.net/specs/oauth-v2-multiple-response-types-1_0.html#Terminology

**Multiple-Valued Response Types**
>The OAuth 2.0 specification allows for registration of space-separated response_type parameter values. If a Response Type contains one of more space characters (%20), it is compared as a space-delimited list of values in which the order of values does not matter.

Comment (joshcanhelp) [reply README.md]: @lbalmaceda - Thank you for that ref! Makes more sense that way. Note to self to correct that in another SDK! :)

**Comment by lbalmaceda** [comment]
LGTM!

File: src/main/java/com/auth0/IdTokenVerifier.java [hunk 24]
```diff
@@ -0,0 +1,166 @@
...
+            }
+            if (!verifyOptions.nonce.equals(nonceClaim)) {
+                throw new TokenValidationException(String.format("Nonce (nonce) claim mismatch in the ID token; expected \"%s\", found \"%s\"", verifyOptions.nonce, nonceClaim));
```
Comment (MattMarji) [inline src/main/java/com/auth0/IdTokenVerifier.java]: String injection / buffer overflow might be possible here if we just spit out the strings. Do we have to be that verbose?

File: src/main/java/com/auth0/IdTokenVerifier.java [hunk 25]
```diff
@@ -0,0 +1,166 @@
...
+            }
+        }
+
+        if (audience.size() > 1) {
+            String azpClaim = decoded.getClaim(AZP_CLAIM).asString();
+            if (isEmpty(azpClaim)) {
+                throw new TokenValidationException("Authorized Party (azp) claim must be a string present in the ID token when Audience (aud) claim has multiple values");
+            }
+            if (!verifyOptions.audience.equals(azpClaim)) {
+                throw new TokenValidationException(String.format("Authorized Party (azp) claim mismatch in the ID token; expected \"%s\", found \"%s\"", verifyOptions.audience, azpClaim));
```
[the remark this answers is not in the document]
Comment (damieng) [reply src/main/java/com/auth0/IdTokenVerifier.java]: We're doing this to help devs understand when things go wrong.  These outputs are exception messages and would not be shown to end users so not really an attack vector for string injection.

You'd have to save the exceptions generated by one user and present them to others and fail to encode that output to make this a vulnerability.

File: src/main/java/com/auth0/RequestProcessor.java [hunk 26]
```diff
@@ -199,34 +222,39 @@ private Tokens exchangeCodeForTokens(String authorizationCode, String redirectUr
...
+            accessToken = codeExchangeTokens.getAccessToken();
+            type = codeExchangeTokens.getType();
+            expiresIn = codeExchangeTokens.getExpiresIn();
+        } else {
+            accessToken = frontChannelTokens.getAccessToken();
+            type = frontChannelTokens.getType();
+            expiresIn = frontChannelTokens.getExpiresIn();
+        }
+
+        // Prefer ID token from the front-channel
```
Comment (MattMarji) [inline src/main/java/com/auth0/RequestProcessor.java]: Why do we prefer ID token from the front-channel?

File: src/test/java/com/auth0/SignatureVerifierTest.java [hunk 27]
```diff
@@ -0,0 +1,193 @@
...
+
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class SignatureVerifierTest {
+
+    private static final String EXPIRED_HS_JWT = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJub25jZSI6IjEyMzQiLCJpc3MiOiJodHRwczovL21lLmF1dGgwLmNvbS8iLCJhdWQiOiJkYU9nbkdzUlloa3d1NjIxdmYiLCJzdWIiOiJhdXRoMHx1c2VyMTIzIiwiZXhwIjo5NzE3ODkzMTd9.5_VOXBmOVMSi8OGgonyfyiJSq3A03PwOEuZlPD-Gxik";
+    private static final String NONE_JWT = "eyJhbGciOiJub25lIiwidHlwIjoiSldUIn0.eyJub25jZSI6IjEyMzQiLCJpc3MiOiJodHRwczovL21lLmF1dGgwLmNvbS8iLCJhdWQiOiJkYU9nbkdzUlloa3d1NjIxdmYiLCJzdWIiOiJhdXRoMHx1c2VyMTIzIn0.";
```
Comment (MattMarji) [inline src/test/java/com/auth0/SignatureVerifierTest.java]: Can we add test(s) for lowercase/uppercase HS256/RS256 and none with mixed upper/lower ie. (None)
This has burned us in the past so i just want to be sure we are covering all cases.

File: src/test/java/com/auth0/InvalidRequestExceptionMatcher.java [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+            return expectedDescription.equals(description);
+        }
+        if (expectedCode != null) {
+            return expectedCode.equals(code);
+        }
+        return false;
+    }
+
+    @Override
+    public void describeTo(Description description) {
```
Comment (MattMarji) [inline src/test/java/com/auth0/InvalidRequestExceptionMatcher.java]: This doesn't seem to ever be used. Is there a reason we are exposing it?

File: src/test/java/com/auth0/InvalidRequestExceptionMatcher.java [hunk 29]
```diff
@@ -0,0 +1,47 @@
+package com.auth0;
+
+import org.hamcrest.Description;
+import org.hamcrest.TypeSafeMatcher;
+
+public class InvalidRequestExceptionMatcher extends TypeSafeMatcher<InvalidRequestException> {
```
Comment (MattMarji) [inline src/test/java/com/auth0/InvalidRequestExceptionMatcher.java]: We don't seem to be using this class at all.
Is there a plan to use it in the future or is there something I may have missed?

File: src/main/java/com/auth0/IdTokenVerifier.java [hunk 30]
```diff
@@ -0,0 +1,166 @@
...
+        Date iatDate = cal.getTime();
+
+        if (now.before(iatDate)) {
+            throw new TokenValidationException(String.format("Issued At (iat) claim error in the ID token; current time (%d) is before issued at time (%d)", now.getTime() / 1000, iatDate.getTime() / 1000));
+        }
+
+
+        if (verifyOptions.nonce != null) {
+            String nonceClaim = decoded.getClaim(NONCE_CLAIM).asString();
+            if (isEmpty(nonceClaim)) {
```
Comment (MattMarji) [inline src/main/java/com/auth0/IdTokenVerifier.java]: is there a reason that we dont check for empty or null nonce first before we assign it to a string?

File: src/main/java/com/auth0/IdTokenVerifier.java [hunk 31]
```diff
@@ -0,0 +1,166 @@
...
+                throw new TokenValidationException("Nonce (nonce) claim must be a string present in the ID token");
+            }
+            if (!verifyOptions.nonce.equals(nonceClaim)) {
+                throw new TokenValidationException(String.format("Nonce (nonce) claim mismatch in the ID token; expected \"%s\", found \"%s\"", verifyOptions.nonce, nonceClaim));
+            }
+        }
+
+        if (audience.size() > 1) {
```
Comment (MattMarji) [inline src/main/java/com/auth0/IdTokenVerifier.java]: why are we using size for audience which is a string?

File: src/test/java/com/auth0/InvalidRequestExceptionMatcher.java (hunk 29 above)
Comment (lbalmaceda) [reply src/test/java/com/auth0/InvalidRequestExceptionMatcher.java]: This is a test helper class. Not shipped in the release jar.

File: src/main/java/com/auth0/IdTokenVerifier.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (MattMarji) [reply src/main/java/com/auth0/IdTokenVerifier.java]: Great, thanks for the context.

File: src/main/java/com/auth0/IdTokenVerifier.java (hunk 31 above)
Comment (gkwang) [inline src/main/java/com/auth0/IdTokenVerifier.java]: As a nit, I'd do audience related checks in the same place so it's easier to track. 
I know the other check is done before this block.

Comment (gkwang) [reply src/main/java/com/auth0/IdTokenVerifier.java]: Actually it's not a problem after some thought. 
The check's purpose is different.
